ddmrp/models/stock_move.py

```diff
@@ -21,3 +21,48 @@ def _merge_moves_fields(self):
         res = super(StockMove, self)._merge_moves_fields()
         res["buffer_ids"] = [(4, m.id) for m in self.mapped("buffer_ids")]
         return res
+
+    def write(self, vals):
+        res = super(StockMove, self).write(vals)
+        if "state" in vals and self.env.company.ddmrp_auto_update_nfp:
+            self._update_ddmrp_nfp()
+        return res
+
+    def _update_ddmrp_nfp(self):
+        if self.env.context.get("no_ddmrp_auto_update_nfp"):
+            return True
+        buffer_model = self.env["stock.buffer"]
+        # Outgoing
+        out_buffers = self.env["stock.buffer"]
+        for move in self:
+            candidates = buffer_model.search(
+                [
+                    ("product_id", "=", move.product_id.id),
+                    ("location_id", "parent_of", move.location_id.id),
+                ]
+            )
+            for cand in candidates:
+                locations = self.env["stock.location"].search(
+                    [("id", "child_of", [cand.location_id.id])]
+                )
+                if move.location_dest_id not in locations:
+                    out_buffers += cand
```

Comment:
The whole computation of in_buffers and out_buffers is going to be slow. It generates N moves = N moves * M buffers sql queries.

It might be better to use less `search()` and use relational fields to leverage the prefetch + compare the parent_path for the location tree.

I used this helper on StockLocation in several modules:

```python
    def is_sublocation_of(self, others):
        """Return True if self is a sublocation of at least one other

        It is equivalent to the "child_of" operator, so it includes itself.
        """
        self.ensure_one()
        # Efficient way to verify that the current location is
        # below one of the other location without using SQL.
        return any(self.parent_path.startswith(other.parent_path) for other in others)

```

Example:

```python

for move in self:
    # with a relation, it will do a single sql query to get all the candidates buffers first, instead of one query per move, if we keep most of the buffers as candidates, it should be a clear win
    candidates = move.mapped("product_id.buffer_ids").filtered(lambda buffer: move.location_id.is_sublocation_of(buffer.location_id))
    for cand in candidates:
        if not move.location_dest_id.is_sublocation_of(cand.location_id):
            out_buffers += cand
```


Also, you should use sets to store the ids before passing everything to a recordset, because continuously updating a recordet is slow (https://gist.github.com/guewen/a4a2eba5d128b4ed3129866f80571fbd)

Reply:
Thanks for the insights! I will test and update now.

Side note, have you thought about proposing the `is_sublocation_of` helper to Odoo?

Reply:
Changes done! could you please have a look?

Reply:
Nice result with the additional improvement adding the second condition in `filtered`, good catch :)



> Side note, have you thought about proposing the `is_sublocation_of` helper to Odoo?

I didn't, I might propose a more generic `is_child_of()` on master (no way they would merge in 13.0), since it makes sense on any hierarchical record.

Reply:
@LoisRForgeFlow https://github.com/odoo/odoo/pull/53866

Reply:
Great! love it :D